test_conformance/api/test_kernel_attributes.cpp

```diff
@@ -0,0 +1,343 @@
+    }
+}
+
+static const std::vector<AttributePermutations*>
+generate_attribute_tests(const KernelAttributes& vec_type_hint_data,
+                         const KernelAttributes& work_group_size_hint_data,
+                         const KernelAttributes& reqd_work_group_size_data)
+{
+    populate_single_attribute_tests(vec_type_hint_data,
+                                    work_group_size_hint_data,
+                                    reqd_work_group_size_data);
+    populate_double_attribute_tests(vec_type_hint_data,
+                                    work_group_size_hint_data,
+                                    reqd_work_group_size_data);
+    populate_triple_attribute_tests(vec_type_hint_data,
+                                    work_group_size_hint_data,
+                                    reqd_work_group_size_data);
+
+    /* Store all of the filled vectors in a single structure */
+    const std::vector<AttributePermutations*> all_tests = {
+        &vect_tests,           &work_tests,           &reqd_tests,
+
+        &work_vect_tests,      &vect_reqd_tests,      &reqd_work_tests,
+
+        &vect_work_reqd_tests, &work_reqd_vect_tests, &reqd_vect_work_tests
+    };
+    return all_tests;
+}
+
+static const std::vector<AttributePermutations*>
+initialise_attribute_data(cl_device_id deviceID)
+{
+    /* This vector stores different work group dimensions that can be used by
+     * the reqd_work_group_size and work_group_size_hint attributes. It
+     * currently only has a single value to minimise time complexity of the
+     * overall test but can be easily changed. */
+    static const std::vector<WorkGroupDimensions> work_group_dimensions = {
+        { 1, 1, 1 }
+    };
+    KernelAttributes vec_type_hint_data = generate_vec_type_hint_data(deviceID);
+    KernelAttributes work_group_size_hint_data =
+        generate_work_group_size_data(work_group_dimensions);
+    KernelAttributes reqd_work_group_size_data =
+        generate_reqd_work_group_size_data(work_group_dimensions);
+
+    /* Generate all the permutations of attributes to create different test
+     * suites */
+    return generate_attribute_tests(vec_type_hint_data,
+                                    work_group_size_hint_data,
+                                    reqd_work_group_size_data);
+}
+
+static bool run_test(cl_context context, cl_device_id deviceID,
+                     const AttributePermutations& permutations)
+{
+    bool success = true;
+    for (auto attribute_permutation : permutations)
+    {
+
+        std::string kernel_source_string =
+            generate_kernel_source(attribute_permutation);
+        const char* kernel_src = kernel_source_string.c_str();
+        clProgramWrapper program;
+        clKernelWrapper kernel;
+        cl_int err = create_single_kernel_helper(context, &program, &kernel, 1,
+                                                 &kernel_src, "test_kernel");
+        ASSERT_SUCCESS(err, "create_single_kernel_helper");
```

Comment:
I think the checks here should use `test_error` instead of `ASSERT_SUCCESS`. The latter will throw an exception and abort the run, which doesn't play nicely when not running this test in isolation.